### PR TITLE
GUI/Recording - Add menu items for recording controls with keyboard shortcuts.

### PR DESCRIPTION
**Pull request by xTVaser**
This makes the recording tools a little more intuitive to use since there still is no proper documentation around it.

In the process, I found out that the current way of adding the keyboard shortcuts to the menu items would omit modifiers such as `shift`.  So I fixed that in both places.

By default wxWidget's `ToString` will return the keyboard character in it's correct case.  For example, if the bind is `Shift+p` it will be printed as such.  However, in most UI's I've seen, the key is capitalized anyway.  So I had to write a small routine to do so.

New menu options (note that the keybinds will only appear once the GS window has been opened, as this is when the majority of pcsx2's keybindings are assigned)
image

Existing binds are still set properly after my changes:
image

### CODE REVIEW

**Comment by RedDevilus** [comment]
What's the purpose of Toggle Recording Mode?

**Comment by xTVaser** [comment]
> What's the purpose of Toggle Recording Mode?

There's currently two recording "modes":
- Recording:
  - new frame's inputs will be saved to the recording file
  - existing frame's inputs will be overwritten
- Replaying:
  - Existing inputs are not modified
  - No new frames will be recorded and will pause at the end of the recording

**Comment by RedDevilus** [comment]
Recording as in it will save to a MP4?
Replaying as in blockdumps or macros?

**Comment by xTVaser** [comment]
None of the above.  The recording tools in pcsx2 record the controller's inputs (an array of bytes) to a file every frame.  When the time comes that you want to export it as an actual video, you would replay it and turn on pcsx2's video dumping feature (F12...if I recall correctly).

File: pcsx2/gui/MainFrame.cpp [hunk 1]
```diff
@@ -765,24 +776,32 @@ void MainEmuFrame::CommitPreset_noTrigger()
...
+void MainEmuFrame::initializeRecordingMenuItem(MenuIdentifiers menuId, wxString keyCodeStr) {
+	wxMenuItem& item = *m_menuRecording.FindChildItem(menuId);
+	wxString text = item.GetItemLabel();
+	const size_t tabPos = text.rfind(L'\t');
+	item.SetItemLabel(text.Mid(0, tabPos ) + L"\t" + keyCodeStr);
+	item.Enable(true);
+}
+void MainEmuFrame::enableRecordingMenuItem(MenuIdentifiers menuId, bool enable)
+{
+	wxMenuItem& item = *m_menuRecording.FindChildItem(menuId);
```
Comment (sonicfind) [inline pcsx2/gui/MainFrame.cpp:801]: The debugger threw a read access violation exception at this point - showing that `m_menuRecording.FindChildItem(menuId)` is already nullptr. This error will appear when you exit the emulator so long as you have booted an ISO at least one time - regardless of if recording tools are enabled or not.

Comment (xTVaser) [reply pcsx2/gui/MainFrame.cpp:801]: Ah ok, that might be due to the recent GUI refactor, I recall you finding an issue there as well?  I can solve that here at the same time...hopefully.  Looking into it.

Comment (xTVaser) [reply pcsx2/gui/MainFrame.cpp:798]: Should be resolved with the latest commit.  The `s*` macros automatically handle the `null` case for windows.

**Comment by sonicfind** [comment]
I noticed that the menu options aren't actually getting disabled on pause or shutdown.
They are however getting enabled on resume or boot. So I went to see why this was. Two big things:
1. The piece that disables the options should be in `Pcsx2App::CloseGsPanel()` instead of `Pcsx2App::OnGsFrameClosed( wxWindowID id )` since the latter is only called on exit... Or at least, I would say put it in `Pcsx2App::CloseGsPanel()`, but
2. `Pcsx2App::CloseGsPanel()` is never actually getting called from anywhere - all thanks to this value in App.h:
```
// Forces the Interface to destroy the GS viewport window when the GS plugin is
// destroyed.  This has the side effect of forcing all plugins to close and re-open
// along with the GS, since the GS viewport window handle will have changed.
static const bool CloseViewportWithPlugins = false;
```
The only time it would get called (AppCorePlugins.cpp) gets short-circuited to false
```
// Yay, this plugin is guaranteed to always be opened first and closed last.
void AppCorePlugins::ClosePlugin_GS()
{
	_parent::ClosePlugin_GS();
	if( CloseViewportWithPlugins && GetMTGS().IsSelf() && GSopen2 ) sApp.CloseGsPanel();
}
```

**Review by RedDevilus — APPROVED** [review]

**Comment by sonicfind** [comment]
768p100%dpi
image
Should be gucci once the small bar on the right side of this is fixed. I see literally nothing else to worry about after that.

**Comment by xTVaser** [comment]
Ran locally.  Shortcuts still showing up:
image
image

lgtm

File: pcsx2/gui/MainFrame.cpp [hunk 2]
```diff
@@ -776,24 +783,32 @@ void MainEmuFrame::CommitPreset_noTrigger()
...
 void MainEmuFrame::AppendKeycodeNamesToMenuOptions() {
-	AppendShortcutToMenuOption( *m_menuSys.FindChildItem( MenuId_Sys_LoadStates ), "States_DefrostCurrentSlot" );
-	AppendShortcutToMenuOption( *m_menuSys.FindChildItem( MenuId_Sys_SaveStates ), "States_FreezeCurrentSlot" );
+	
+	AppendShortcutToMenuOption(*m_menuSys.FindChildItem( MenuId_Sys_LoadStates ), wxGetApp().GlobalAccels->findKeycodeWithCommandId("States_DefrostCurrentSlot").toTitleizedString());
+	AppendShortcutToMenuOption(*m_menuSys.FindChildItem( MenuId_Sys_SaveStates ), wxGetApp().GlobalAccels->findKeycodeWithCommandId("States_FreezeCurrentSlot").toTitleizedString());
 }
 
+#ifndef DISABLE_RECORDING
+void MainEmuFrame::initializeRecordingMenuItem(MenuIdentifiers menuId, wxString keyCodeStr, bool enable) {
```
Comment (refractionpcsx2) [inline pcsx2/gui/MainFrame.cpp:799]: Bracket needs to go on to the next line

File: pcsx2/gui/MainFrame.cpp [hunk 3]
```diff
@@ -776,24 +783,32 @@ void MainEmuFrame::CommitPreset_noTrigger()
...
+	wxMenuItem& item = *m_menuRecording.FindChildItem(menuId);
+	wxString text = item.GetItemLabel();
+	const size_t tabPos = text.rfind(L'\t');
+	item.SetItemLabel(text.Mid(0, tabPos ) + L"\t" + keyCodeStr);
+	item.Enable(enable);
+}
+void MainEmuFrame::enableRecordingMenuItem(MenuIdentifiers menuId, bool enable)
```
Comment (refractionpcsx2) [inline pcsx2/gui/MainFrame.cpp:806]: maybe add a space between functions

**Review by refractionpcsx2 — CHANGES_REQUESTED** [review]

**Review by refractionpcsx2 — APPROVED** [review]

**Pull request merged**